[WIP] Update org settings to differentiate between full members and members.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
#14409 

**Testing Plan:** <!-- How have you tested? -->


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
The first commit looks good; though I'd use `models:` as the prefix.

For the second commit, I think the `ALLOWED_TO` names are awkward.  I could consider doing names like `POLICY_MEMBERS_ONLY` for that concept.

But I'm also not sure that the common variables in the second commit are next helpful; I'll need to think about it more.

timabbott commented:
I think probably the second commit's approach is net helpful; it'll be useful for features like `add_emoji_by_admins_only` that we do plan to have a simple "what set of users can do this" model and it'll be helpful to have common fields for the ones that are simple (bot_creation_policy is not).

sahil839 commented:
@timabbott I have made the change in first commit message and have also changed the variable names in second commit.
I agree that it may be not be useful for all the settings to have common values but it may be helpful for some of the some settings that will have same choices.

timabbott commented:
I merged the first commit.  For the second, can you rename PERMISSION_TYPES to `COMMON_PERMISSION_TYPES`?  I think that feels more readable.

timabbott commented:
I merged the second commit as b66a67272f6c294f65f860e8303aa12fe104eede, after renaming that variable to `COMMON_POLICY_TYPES`, since that felt more readable to me.

timabbott commented:
For the last commit, I don't like the readability of `can_perform_action("create_stream_policy)`.  How about `user_profile.has_permission("create_stream_policy")`?

I'm not sure the rewrite in `events.py` improves readability.

sahil839 commented:
I have made the change of can_perform_action to has_permission.
Regarding the change of events.py, I have made these changes as after we will add the new members to more settings, this will avoid duplication of code, but yeah I agree that the previous code had better readability.
Should I remove the change in events.py for now and start working on adding the new member option and then may be we can figure out the way to avoid duplication and have nice readability.

sahil839 commented:
@timabbott I have renamed can_confirm_action to has_permission. 
I have also changed variable names in events.py for more readability.

Maybe we can just leave the events.py change for now and start working on adding new members option to settings one by one and while doing it I can look at other ways to avoid duplication in events.py.

timabbott commented on zerver/lib/events.py:
```diff
@@ -528,14 +528,23 @@ def apply_event(state: Dict[str, Any],
-                state['can_create_streams'] = user_profile.can_create_streams()
+            permission_types = ['can_create_streams', 'can_subscribe_other_users']
+            policy_types = ['create_stream_policy', 'invite_to_stream_policy']
+            policy_permission_dict = {policy_types[i]: permission_types[i] for i in range(len(policy_types))}
```
This code is weird and hard to read -- why not just make create a hardcoded dictionary rather than doing this weird zipping of two lists to make it?

timabbott commented on zerver/lib/streams.py:
```diff
@@ -281,7 +281,7 @@ def list_to_streams(streams_raw: Iterable[Mapping[str, Any]],
     else:
         # autocreate=True path starts here
-        if not user_profile.can_create_streams():
+        if not user_profile.has_permission('create_stream_policy'):
```
I think for readability, it might be better to keep `.can_create_streams()` and friends as wrappers around `.has_permission`; the old code is a lot more readable in most places where we don't have a dynamic setting.

sahil839 replied:
@timabbott do you mean something like this-
```
def has_permission(self, policy_name: str) -> bool:
        
        if policy_name == 'create_stream_policy':
            return self.can_create_streams()

        if policy_name == 'invite_to_stream_policy':
            return self.can_subscribe_other_users()

        raise AssertionError("Invalid policy")
```
and `can_create_streams` ans similar methods remain uncahnged.

sahil839 commented on zerver/tests/test_subs.py:
```diff
@@ -2269,40 +2269,40 @@ def test_subscriptions_add_stream_with_null(self) -> None:
         othello = self.example_user('othello')
         othello.role = UserProfile.ROLE_REALM_ADMINISTRATOR
-        self.assertTrue(othello.can_create_streams())
+        self.assertTrue(othello.has_permission('create_stream_policy'))
```
@timabbott are tests fine using has_permission or I should keep `can_create_streams` as you suggested above to keep it as wrappers?

timabbott replied:
I think `can_create_streams` might be more readable here, but either way is probably fine in tests.

sahil839 commented:
@timabbott I have made changes.

timabbott commented on zerver/models.py:
```diff
@@ -1165,6 +1165,16 @@ def can_subscribe_other_users(self) -> bool:
+        if policy_name == 'invite_to_stream_policy':
+            return self.can_subscribe_other_users()
+
+        raise AssertionError("Invalid policy")
```
Sorry, this isn't what I had in mind -- what I meant is to create a common `has_permission` function, and make `self.can_create_streams()` call that function with `create_stream_policy` as the argument,

timabbott commented on zerver/lib/events.py:
```diff
@@ -528,14 +528,23 @@ def apply_event(state: Dict[str, Any],
-            if (field in ['realm_invite_to_stream_policy',
-                          'realm_waiting_period_threshold']) and 'can_subscribe_other_users' in state:
-                state['can_subscribe_other_users'] = user_profile.can_subscribe_other_users()
+            policy_types = ['create_stream_policy', 'invite_to_stream_policy']
```
I think you don't need this variable; everywhere you use it, you can use `policy_permission_dict` instead.

sahil839 commented:
@timabbott I have made the changes as told by you.

sahil839 commented:
@timabbott any more changes needed?

timabbott commented:
Yeah, this finally looks good, merged, thanks @sahil839!
